core/src/main/java/io/questdb/cairo/AbstractIndexReader.java

```diff
@@ -164,6 +153,43 @@ public void of(CairoConfiguration configuration, Path path, CharSequence name, l
         }
     }
 
+    private void readIndexMetadataAtomically() {
+        // Triple check atomic read. We read first and last sequences. If they match - there is a chance at stable
+        // read. Confirm start sequence hasn't changed after values read. If it has changed - retry the whole thing.
+        final long deadline = clock.getTicks() + spinLockTimeoutMs;
+        while (true) {
+            long seq = keyMem.getLong(BitmapIndexUtils.KEY_RESERVED_OFFSET_SEQUENCE);
+            int keyCount;
+            long valueMemSize;
+            int blockValueCountMod;
+
+            Unsafe.getUnsafe().loadFence();
+            if (keyMem.getLong(BitmapIndexUtils.KEY_RESERVED_OFFSET_SEQUENCE_CHECK) == seq) {
+                blockValueCountMod = keyMem.getInt(BitmapIndexUtils.KEY_RESERVED_OFFSET_BLOCK_VALUE_COUNT) - 1;
+                keyCount = keyMem.getInt(BitmapIndexUtils.KEY_RESERVED_OFFSET_KEY_COUNT);
+                valueMemSize = keyMem.getLong(BitmapIndexUtils.KEY_RESERVED_OFFSET_VALUE_MEM_SIZE);
+
+                Unsafe.getUnsafe().loadFence();
+                if (keyMem.getLong(BitmapIndexUtils.KEY_RESERVED_OFFSET_SEQUENCE) == seq) {
+                    this.keyFileSequence = seq;
+                    this.valueMemSize = valueMemSize;
+                    this.keyCount = keyCount;
+                    this.blockValueCountMod = blockValueCountMod;
+                    this.blockCapacity = (blockValueCountMod + 1) * 8 + BitmapIndexUtils.VALUE_BLOCK_FILE_RESERVED;
+                    this.keyCountIncludingNulls = columnTop > 0 ? keyCount + 1 : keyCount;
+                    break;
+                }
+            }
+
+            if (clock.getTicks() > deadline) {
+                LOG.error().$(INDEX_CORRUPT).$(" [timeout=").$(spinLockTimeoutMs).$("ms]").$();
+                throw CairoException.critical(0).put(INDEX_CORRUPT);
+            }
+
+            Os.pause();
+        }
```

Comment:
nit: `return valueMemSize;` here instead of using the field. It would be a better code readablity